ppl/lake/tsdirstream.go

```diff
@@ -0,0 +1,81 @@
+package lake
+
+import (
+	"context"
+	"sort"
+
+	"github.com/brimsec/zq/ppl/lake/chunk"
+	"github.com/brimsec/zq/zbuf"
+	"golang.org/x/sync/errgroup"
+)
+
+type tsDirStream struct {
+	ch  chan tsDirStreamResult
+	err error
+}
+
+type tsDirStreamResult struct {
+	tsDir  tsDir
+	chunks []chunk.Chunk
+}
+
+func newTsDirStream(ctx context.Context, lk *Lake, tsDirs []tsDir) *tsDirStream {
+	sort.Slice(tsDirs, func(i, j int) bool {
+		if lk.DataOrder == zbuf.OrderAsc {
+			return tsDirs[i].Ts < tsDirs[j].Ts
+		}
+		return tsDirs[j].Ts < tsDirs[i].Ts
+	})
+
+	t := &tsDirStream{ch: make(chan tsDirStreamResult)}
+	go t.run(ctx, lk, tsDirs)
+	return t
+}
+
+func (t *tsDirStream) run(ctx context.Context, lk *Lake, tsDirs []tsDir) {
+	// Use a channel of channels to ensure results are returned in the correct
+	// order.
+	tsDirChs := make(chan chan tsDirStreamResult, len(tsDirs))
```

Comment:
Could this be a `[]chan tsDirStreamResult` instead of a channel of channels?

Reply:
Yes actually and I think this is better.

Reply:
It's certainly possible but requires explicit synchronization. I think the channel of channels winds up being simpler as a result, but to be clear, I'm fine with either approach.

Reply:
One benefit of using the chan of chans I suppose is that it could be used to limit the number of in flight tsdir fetches at a given time. That said we're not currently doing this.

Reply:
> It's certainly possible but requires explicit synchronization.

@nwt I think I am missing something. How so?

Reply:
The list of tsDirChans would be fully populated in the main goroutine before being iterated on in the consumer goroutine.

Reply:
Ah, in that case, no synchronization required.